crates/tower-cmd/src/mcp.rs

```diff
@@ -0,0 +1,512 @@
+}
+
+#[derive(Debug, Deserialize, JsonSchema)]
+struct AddParameterRequest {
+    #[serde(flatten)]
+    common: CommonParams,
+    name: String,
+    description: String,
+    default: String,
+}
+
+#[derive(Debug, Deserialize, JsonSchema)]
+struct GenerateTowerfileRequest {
+    #[serde(flatten)]
+    common: CommonParams,
+    script_path: Option<String>,
+}
+
+#[derive(Debug, Deserialize, JsonSchema)]
+struct EmptyRequest {
+    #[serde(flatten)]
+    common: CommonParams,
+}
+
+
+pub fn mcp_cmd() -> Command {
+    Command::new("mcp-server")
+        .about("Runs a local SSE MCP server for LLM interaction")
+        .arg(
+            clap::Arg::new("port")
+                .long("port")
+                .short('p')
+                .help("Port for SSE server (default: 34567)")
+                .default_value("34567")
+                .value_parser(clap::value_parser!(u16)),
+        )
+}
+
+pub async fn do_mcp_server(config: Config, args: &clap::ArgMatches) -> Result<()> {
+    let port = *args.get_one::<u16>("port").unwrap();
+    let bind_addr = format!("127.0.0.1:{}", port);
+    
+    eprintln!("SSE server running on http://{}", bind_addr);
+    
+    let ct = SseServer::serve(bind_addr.parse()?)
+        .await?
+        .with_service_directly(move || TowerService::new(config.clone()));
+
+    tokio::signal::ctrl_c().await?;
+    ct.cancel();
+    Ok(())
+}
+
+#[derive(Clone)]
+pub struct TowerService {
+    config: Config,
+    tool_router: ToolRouter<Self>,
+}
+
+#[tool_router]
+impl TowerService {
+    pub fn new(config: Config) -> Self {
+        Self {
+            config: std::env::var("TOWER_JWT")
+                .ok()
+                .and_then(|token| Session::from_jwt(&token).ok())
+                .map(|session| config.clone().with_session(session))
+                .unwrap_or(config),
+            tool_router: Self::tool_router(),
+        }
+    }
+
+    fn json_success<T: serde::Serialize>(data: T) -> Result<CallToolResult, McpError> {
+        let text = serde_json::to_string_pretty(&data)
+            .map_err(|e| McpError::internal_error("Serialization failed", Some(json!({"error": e.to_string()}))))?;
+        Ok(CallToolResult::success(vec![Content::text(text)]))
+    }
+
+    fn text_success(message: String) -> Result<CallToolResult, McpError> {
+        Ok(CallToolResult::success(vec![Content::text(message)]))
+    }
+
+    fn error_result(prefix: &str, error: impl std::fmt::Display + std::fmt::Debug) -> Result<CallToolResult, McpError> {
+        Ok(CallToolResult::error(vec![Content::text(format!("{}: {:#?}", prefix, error))]))
+    }
+    
+    fn resolve_working_directory(common: &CommonParams) -> std::path::PathBuf {
+        common.working_directory
+            .as_ref()
+            .map(std::path::PathBuf::from)
+            .unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| std::path::PathBuf::from(".")))
+    }
+
+
+    #[tool(description = "List all Tower apps in your account")]
```

Comment:
The rust MCP library, `rmcp`, has this nice abstraction where you annotate async functions (in a `tool_router` struct implementation) with the description of the tool. It then picks out them all and generates the endpoints and everything needed for the MCP server.